### ISSUE TITLE
Update Unicorn Engine to > 2.1.1

### ISSUE THREAD

**Issue by disconnect3d** [pwndbg/pwndbg#2588]
When UE is updated to > 2.1.1 we should revert the if/else for arm64 bug that we applied a workaround for in here: https://github.com/pwndbg/pwndbg/issues/2548

**Comment by patryk4815** [comment]
Btw unicorn v2.1.0 introduced:
`Now Unicorn no longer allocates 2GB memory for every instance. The memory will be only committed once used and the upper limit can be adjusted with uc_ctl.`


```
def get_buf_size(emu):
    carg = ctypes.c_uint32()
    emu.uc.ctl(U.UC_CTL_TCG_BUFFER_SIZE, U.UC_CTL_IO_READ, ctypes.byref(carg))
    return carg.value

def set_buf_size(emu):
    emu.uc.ctl(U.UC_CTL_TCG_BUFFER_SIZE, U.UC_CTL_IO_WRITE, ctypes.c_uint32(50*1024*1024))
```

Do we need still this code?
https://github.com/pwndbg/pwndbg/blob/1980aba40d6303d23234e09a76cf59625beec2f0/pwndbg/aglib/disasm/__init__.py#L302-L328

**Comment by disconnect3d** [comment]
I've got `unicorn==2.1.1` but I still get `Could not allocate dynamic translator buffer` on my VPS with 500MB RAM, so I guess no?

**Comment by wtdcode** [comment]
> I've got `unicorn==2.1.1` but I still get `Could not allocate dynamic translator buffer` on my VPS with 500MB RAM, so I guess no?

Hmmm, is it Linux or Windows?

**Comment by patryk4815** [comment]
@wtdcode linux x86

**Comment by wtdcode** [comment]
> @wtdcode linux x86

That’s weird. Do you enable overcommitting?

**Comment by patryk4815** [comment]
@disconnect3d https://github.com/pwndbg/pwndbg/blob/1980aba40d6303d23234e09a76cf59625beec2f0/pwndbg/emu/emulator.py#L194-L195
this issue probably is fixed, we need check 🙃

**Comment by wtdcode** [comment]
> I've got `unicorn==2.1.1` but I still get `Could not allocate dynamic translator buffer` on my VPS with 500MB RAM, so I guess no?

By the way, we have `UC_CTL_TCG_BUFFER_SIZE` to determine the size of dynamic buffer Unicorn allocates.

**Comment by disconnect3d** [comment]
It's a x86-64 vps on DO with 1GB ram (my bad, thought its 500MB). I believe the memory overcommiting is disabled:
 
```
# uname -a
Linux pwndbg 6.5.0-9-generic #9-Ubuntu SMP PREEMPT_DYNAMIC Sat Oct  7 01:35:40 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

# sysctl -a | rg overcommit
vm.nr_overcommit_hugepages = 0
vm.overcommit_kbytes = 0
vm.overcommit_memory = 0
vm.overcommit_ratio = 50
```

For what is worth, I added this to our emulator construction as per @patryk4815 suggestion:
```py
self.uc.ctl(U.UC_CTL_TCG_BUFFER_SIZE, U.UC_CTL_IO_WRITE, ctypes.c_uint32(1*1024*1024))
```
and it works!

**Comment by disconnect3d** [comment]
@wtdcode Maybe UE could automagically detect "how much memory can it use" by lowering down its memory setting if the machine doesn't have enough memory and the allocation for the buffer fails?

**Comment by disconnect3d** [comment]
Also just to be sure, how does UE use this buffer size limit? Does it slow down things if we set it to 1MB (I assume it can?)? Does UE ever resize this buffer? What if it fails to? Will it just exit as it does initially?

**Comment by wtdcode** [comment]
I found the culprit is Linux's ["Heuristic overcommit handling"](https://www.kernel.org/doc/Documentation/vm/overcommit-accounting) as suggested by "vm.overcommit_memory = 0" in your config, which I believe most Linux distributions have as default config. Note this actually **enables** the overcommitting. However, Linux still has a upper limit of the memory a process can overcommit and in this case it looks like kernel [deems 2GB as too much memory](https://unix.stackexchange.com/questions/378172/what-does-heuristics-in-overcommit-memory-0-mean) and refuses it. (Like, does something like 1.5G or 1.2GB work?) 

A side note, 2.1.x doesn't have impact on the memory management on *nix, where Unicorn always relies on the overcomitting feature. But `UC_CTL_TCG_BUFFER_SIZE` indeed is the workaround to solve the issue here. (I'm sorry I forget to add a helper function to Python bindings. I will add it in the next release.)

> Maybe UE could automagically detect "how much memory can it use" by lowering down its memory setting if the machine doesn't have enough memory and the allocation for the buffer fails?

Unfortunately, this is hard to achieve because we don't know users want to spin up one instance or multiple instances.

> Also just to be sure, how does UE use this buffer size limit? Does it slow down things if we set it to 1MB (I assume it can?)? Does UE ever resize this buffer? What if it fails to? Will it just exit as it does initially?

Unicorn simply unconditionally allocates this buffer and re-uses it as a memory pool. 1MB would be too small for Unicorn and the minimum memory to run some simple snippets is 32~64MB if I recall it correctly. For pwndbg, I think 512MB is pretty reasonable and enough.

Unicorn never resizes this buffer and will crash if it runs out of memory. Unfortunately, it is not intended by Unicorn and inherited from QEMU. Many QEMU code is rather hard to rework to accept memory allocation failures. Some code even relies this behavior by calling abort directly.

**Comment by wtdcode** [comment]
As a user of pwndbg (yeah, I debug unicorn with pwndbg =p), I think a good approach is keeping the `mmap` detection in the current codebase but doing it by trying mmap 2GB, 1GB, 512MB, 256MB... until a suitable memory size is found. If there is not enough memory, maybe just disable unicorn then. This ensures all users will have the best experience.

**Comment by patryk4815** [comment]
> trying mmap 2GB, 1GB, 512MB, 256MB... until a suitable memory

This is good 👍

**Comment by patryk4815** [comment]
btw @Arusekk @peace-maker ^ you also have this mmap check in pwntools, now it can be improved :)

**Comment by wtdcode** [comment]
Ping here for 2.1.2 =)

**Comment by patryk4815** [comment]
@wtdcode thanks! Now waiting for loongarch support :P

**Comment by wtdcode** [comment]
Hmm, you guys are so fast =). I recall loongarch was supported by gdb just a few months ago? Anyway, it will be supported in the next release 2.2.0.

**Comment by patryk4815** [comment]
@wtdcode looks like unicorn 2.1.2 introduce new regression
It is not possible to build unicorn on armv7l / i686 (all 32bit systems).

```
[ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/cpu-exec-common.c.o
[ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/cputlb.c.o
[ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-all.c.o
[ 78%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-runtime.c.o
[ 78%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-runtime-gvec.c.o
[ 78%] Building C object CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/machine.c.o
[ 78%] Building C object CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/mem_helper.c.o
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c: In function ‘helper_stq_le_parallel’:
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:423:12: error: ‘return’ with a value, in function returning void [-Wreturn-mismatch]
  423 |     return 0;
      |            ^
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:411:6: note: declared here
  411 | void helper_stq_le_parallel(CPUPPCState *env, target_ulong addr,
      |      ^~~~~~~~~~~~~~~~~~~~~~
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c: In function ‘helper_stq_be_parallel’:
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:439:12: error: ‘return’ with a value, in function returning void [-Wreturn-mismatch]
  439 |     return 0;
      |            ^
/build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:427:6: note: declared here
  427 | void helper_stq_be_parallel(CPUPPCState *env, target_ulong addr,
      |      ^~~~~~~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/ppc64-softmmu.dir/build.make:580: CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/mem_helper.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:463: CMakeFiles/ppc64-softmmu.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by wtdcode** [comment]
> [@wtdcode](https://github.com/wtdcode) looks like unicorn 2.1.2 introduce new regression It is not possible to build unicorn on armv7l / i686 (all 32bit systems).
> 
> ```
> [ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/cpu-exec-common.c.o
> [ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/cputlb.c.o
> [ 77%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-all.c.o
> [ 78%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-runtime.c.o
> [ 78%] Building C object CMakeFiles/riscv64-softmmu.dir/qemu/accel/tcg/tcg-runtime-gvec.c.o
> [ 78%] Building C object CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/machine.c.o
> [ 78%] Building C object CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/mem_helper.c.o
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c: In function ‘helper_stq_le_parallel’:
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:423:12: error: ‘return’ with a value, in function returning void [-Wreturn-mismatch]
>   423 |     return 0;
>       |            ^
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:411:6: note: declared here
>   411 | void helper_stq_le_parallel(CPUPPCState *env, target_ulong addr,
>       |      ^~~~~~~~~~~~~~~~~~~~~~
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c: In function ‘helper_stq_be_parallel’:
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:439:12: error: ‘return’ with a value, in function returning void [-Wreturn-mismatch]
>   439 |     return 0;
>       |            ^
> /build/unicorn-2.1.2/src/qemu/target/ppc/mem_helper.c:427:6: note: declared here
>   427 | void helper_stq_be_parallel(CPUPPCState *env, target_ulong addr,
>       |      ^~~~~~~~~~~~~~~~~~~~~~
> make[2]: *** [CMakeFiles/ppc64-softmmu.dir/build.make:580: CMakeFiles/ppc64-softmmu.dir/qemu/target/ppc/mem_helper.c.o] Error 1
> make[1]: *** [CMakeFiles/Makefile2:463: CMakeFiles/ppc64-softmmu.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ```

Hmm, really? We should have CI for these platforms?

**Comment by wtdcode** [comment]
Oh yeah, looks like we don't have CI for any arm 32bit platform. I do think we never really support it. Could you file an issue so I can track this?

**Comment by wtdcode** [comment]
*i686 x86 will work, like this: https://github.com/unicorn-engine/unicorn/actions/runs/13309809343/job/37172815254

**Comment by patryk4815** [comment]
@wtdcode I think supporting armv7l would be nice, there ale a lot of armv7l systems still. In 2.1.1 it compile fine.

**Comment by wtdcode** [comment]
Yeah, the fix is easy from the error message but releasing means another CI workflow to be added, which we don't have at this moment.

**Comment by patryk4815** [comment]
> *i686 x86 will work, like this: https://github.com/unicorn-engine/unicorn/actions/runs/13309809343/job/37172815254

🤔 maybe different compiler? We use gcc-14

**Comment by wtdcode** [comment]
From what logs says [here](https://github.com/unicorn-engine/unicorn/actions/runs/13309809343/job/37172815254#step:9:381), we are using `gcc-10` instead. Maybe that's the culprit?

**Comment by patryk4815** [comment]
unicorn 2.1.1 with gcc-14 compile success 
unicorn 2.1.2 with gcc-14 compile failure 
unicorn 2.1.2 with gcc-13 compile success 
Do you want ticket in your repo?

**Comment by wtdcode** [comment]
> unicorn 2.1.1 with gcc-14 compile success unicorn 2.1.2 with gcc-14 compile failure unicorn 2.1.2 with gcc-13 compile success Do you want ticket in your repo?

Yeah please also attach the full log. I will see if we can add a new workflow for this so that we can have arm32 wheels.
